### PR TITLE
[ADDED] Allow certain consumer attributes to be updated #2670, #2603

### LINKED ISSUE #2670: Allow modification of some JS push consumer settings
The request is for being able to alter the MaxAckPending value for an existing push consumer, as it is possible in STAN.

Use case: when you deploy a larger number of workers than the current MaxAckPending value, you are not making use of all your available workers and you need to adjust that value.

Looking further ahead, it could also be interesting to allow adjusting AckWait, Description, RateLimit and MaxDeliver

### PR DESCRIPTION
**Pull request by derekcollison**
Signed-off-by: Derek Collison <derek@nats.io>

Resolves #2670, #2603

### Changes proposed in this pull request:

/cc @nats-io/core

### CODE REVIEW

File: server/consumer.go [hunk 1]
```diff
@@ -1138,13 +1125,172 @@ func (o *consumer) forceExpirePending() {
...
+// Update the config based on the new config, or error if update not allowed.
+func (o *consumer) updateConfig(cfg *ConsumerConfig) error {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	if err := o.acc.checkNewConsumerConfig(&o.cfg, cfg); err != nil {
+		return err
+	}
+
+	// Something different, so check since we only allow certain things to be updated.
```
Comment (ripienaar) [inline server/consumer.go:1227]: these are duplicated of the other ones? maybe make a single function and reuse? Later easier to maintain just one

**Review by ripienaar — COMMENTED** [review]
seems good other than the non dry code, but I dont know these code paths well

**Comment by derekcollison** [comment]
I am reviewing failing test atm.

File: server/consumer.go (hunk 1 above)
Comment (derekcollison) [reply server/consumer.go:1227]: Oversight, I added the check to remove that code.. ;) Will update.

File: server/jetstream_cluster.go [hunk 2]
```diff
@@ -2763,15 +2771,28 @@ func (js *jetStream) processClusterCreateConsumer(ca *consumerAssignment, state
...
+			js.processConsumerLeaderChange(o, true)
+		} else {
 			if !alreadyRunning {
 				s.startGoRoutine(func() { js.monitorConsumer(o, ca) })
 			}
-		} else {
-			// Single replica consumer, process manually here.
-			js.processConsumerLeaderChange(o, true)
+			// Process if existing.
+			//			if wasExisting && (o.isLeader() || rg.node.GroupLeader() == _EMPTY_) {
```
Comment (kozlovic) [inline server/jetstream_cluster.go:2786]: Remove?

**Review by kozlovic — APPROVED** [review]
LGTM, except for a left over comment..

**Review by kozlovic — APPROVED** [review]
LGTM

**Comment by NoamAP** [comment]
Hi, I'm trying to update an existing consumer config using the api "$JS.API.CONSUMER.CREATE". 
It works great when max consumers > 1. On a stream with max consumers == 1, I get 
```
{"type":"io.nats.jetstream.api.v1.consumer_create_response","error":{"code":400,"err_code":10026,"description":"maximum consumers limit reached"}}
``` 

Is this intended? Any ideas how to update the consumer config in that case?

**Comment by derekcollison** [comment]
Sounds like might be a bug, if you could file an issue against the server and we will get it fixed.

In meantime, you could edit stream to max consumers 2, update consumer, and put max consumers back.

**Comment by scottf** [comment]
@NoamAP Can you give some more details about your server version, which client if any you are using, your full stream configuration, your consumer configuration and the field you are trying to change? I'm trying to replicate the issue like so...

```
MessageToServer -->
  Subject: '$JS.API.STREAM.DELETE.maxstream' 
  Payload: <no data>

MessageToServer -->
  Subject: $JS.API.STREAM.CREATE.maxstream  
  Payload: {"name":"maxstream","subjects":["maxsub"],"retention":"limits","max_consumers":1,"storage":"memory","num_replicas":1,"discard":"old"}

Stream Info Response:
{"did_create":true,"created":"2024-06-02T17:01:34.8233068Z","state":{"first_seq":0,"bytes":0,"first_ts":"0001-01-01T00:00:00Z","messages":0,"last_ts":"0001-01-01T00:00:00Z","consumer_count":0,"last_seq":0},"config":{"discard":"old","duplicate_window":120000000000,"consumer_limits":{},"deny_delete":false,"sealed":false,"subjects":["maxsub"],"max_msg_size":-1,"allow_rollup_hdrs":false,"max_bytes":-1,"storage":"memory","allow_direct":false,"max_age":0,"max_consumers":1,"max_msgs_per_subject":-1,"num_replicas":1,"name":"maxstream","deny_purge":false,"compression":"none","max_msgs":-1,"retention":"limits","mirror_direct":false},"ts":"2024-06-02T17:01:34.8233068Z"}

MessageToServer --> 
  Subject: $JS.API.CONSUMER.CREATE.maxstream.maxcon
  Payload: {"stream_name":"maxstream","config":{"name":"maxcon","deliver_policy":"all","ack_policy":"explicit","replay_policy":"instant"}

Consumer Info Response:
{"num_redelivered":0,"stream_name":"maxstream","created":"2024-06-02T17:01:34.8529751Z","num_waiting":0,"num_pending":0,"name":"maxcon","delivered":{"consumer_seq":0,"stream_seq":0},"num_ack_pending":0,"config":{"max_ack_pending":1000,"max_waiting":512,"num_replicas":0,"name":"maxcon","ack_wait":30000000000,"replay_policy":"instant","max_deliver":-1,"inactive_threshold":5000000000,"deliver_policy":"all","ack_policy":"explicit"},"ack_floor":{"consumer_seq":0,"stream_seq":0},"ts":"2024-06-02T17:01:34.8529751Z"}

MessageToServer --> 
  Subject: $JS.API.CONSUMER.CREATE.maxstream.maxcon
  Payload: {"stream_name":"maxstream","config":{"description":"desc","name":"maxcon","deliver_policy":"all","ack_policy":"explicit","replay_policy":"instant"}}

Consumer Info Response:
{"num_redelivered":0,"stream_name":"maxstream","created":"2024-06-02T17:01:34.8529751Z","num_waiting":0,"num_pending":0,"name":"maxcon","delivered":{"consumer_seq":0,"stream_seq":0},"num_ack_pending":0,"config":{"max_ack_pending":1000,"max_waiting":512,"num_replicas":0,"name":"maxcon","ack_wait":30000000000,"replay_policy":"instant","description":"desc","max_deliver":-1,"inactive_threshold":5000000000,"deliver_policy":"all","ack_policy":"explicit"},"ack_floor":{"consumer_seq":0,"stream_seq":0},"ts":"2024-06-02T17:01:34.8560845Z"}
```

**Comment by derekcollison** [comment]
@scottf If stream's max consumers set to 1, a consumer update will fail.

**Comment by scottf** [comment]
@NoamAP Just wanted to reiterate. I attempted to do what you said. I have a very minimal stream and consumer configuration and I was able to update the consumer fine. It's entirely possible it's specific to the field your are trying to modify, which for any number of reasons may not be modifiable, so I'm just looking for more details so I can reproduce this. If you have a repeatable example or something for me to try, please open an issue in this repo or at least DM me on slack and we'll get this figured out.

**Comment by NoamAP** [comment]
Thank you! I sincerely apologise I didn't send more details yet. I want to upgrade nats and see if it still happens.
I will update here but it will take some time until I get to doing it.

**Comment by NoamAP** [comment]
Opened #5488

**Pull request merged**